File: astar/2019/R1D/1001/Main.java
import java.util.*;
import java.math.*;

public class Main {
    final static BigInteger b7 = BigInteger.valueOf(7); 
    final static BigInteger b18 = BigInteger.valueOf(18); 

    BigInteger solve(long n, BigInteger a, BigInteger b) {
        if (n == 0) return BigInteger.ZERO;
        BigInteger bn = BigInteger.valueOf(n);
        BigInteger res = bn;
        for (int i = 0; i < 2; ++ i) res = res.multiply(bn);
        res = res.multiply(b);
        BigInteger v = BigInteger.valueOf(n - 1);
        for (int i = 0; i < 2; ++ i) v = v.multiply(bn);
        v = v.multiply(a);
        res = res.add(v);
        if ((n & 1) == 0) {
            BigInteger ret = solve(n >> 1, a, b);
            ret = ret.multiply(b7);
            v = BigInteger.valueOf(n >> 1);
            v = v.multiply(v).multiply(b18).multiply(a);
            ret = ret.add(v);
            res = res.min(ret);
        }
        return res;
    }

    public static void main(String[] args) {
        final BigInteger m = BigInteger.valueOf(1000000007);
        Main obj = new Main();
        Scanner s = new Scanner(System.in);
        int t = s.nextInt();
        for (int i = 0; i < t; ++ i) {
            long n = s.nextLong();
            int a = s.nextInt();
            int b = s.nextInt();
            BigInteger res = obj.solve(n, BigInteger.valueOf(a), BigInteger.valueOf(b));
            System.out.println(res.mod(m));
        }
        s.close();
    }
}

File: astar/2019/R1D/1001/main.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long LL;

int solve(int n, int a, int b) {
    if (n == 0) return 0;
    const int mod = 1000000007;
    LL res = n;
    for (int i = 0; i < 2; ++ i) res = (res * n) % mod;
    res = (res * b) % mod;
    LL v = n - 1;
    for (int i = 0; i < 2; ++ i) v = (v * n) % mod;
    v = (v * a) % mod;
    res = (res + v) % mod;
    if (!(n & 1)) {
        LL ret = solve(n >> 1, a, b);
        ret = (ret * 7) % mod;
        v = n >> 1;
        v = (v * v) % mod;
        v = (v * 18) % mod;
        v = (v * a) % mod;
        ret = (ret + v) % mod;
        res = min(res, ret);
    }
    return res;
}

int main(int argc, char** argv) {
    int t;
    cin >> t;
    for (int i = 0; i < t; ++ i) {
        int n, a, b;
        cin >> n >> a >> b;
        auto ret = solve(n, a, b);
        cout << ret << endl;
    }
    return 0;
}
